Do not read a JAR completely into a byte-array before computing the SHA.

### New feature, improvement proposal

In our CI/CD pipeline (Linux, Java 21), we encountered this error:

```
Caused by: java.lang.OutOfMemoryError: Cannot reserve 1490778611 bytes of direct buffer memory (allocated: 1768309554, limit: 3221225472)
    at java.nio.Bits.reserveMemory (Bits.java:178)
    at java.nio.DirectByteBuffer.<init> (DirectByteBuffer.java:127)
    at java.nio.ByteBuffer.allocateDirect (ByteBuffer.java:360)
    at sun.nio.ch.Util.getTemporaryDirectBuffer (Util.java:242)
    at sun.nio.ch.IOUtil.read (IOUtil.java:303)
    at sun.nio.ch.IOUtil.read (IOUtil.java:283)
    at sun.nio.ch.FileChannelImpl.read (FileChannelImpl.java:234)
    at sun.nio.ch.ChannelInputStream.read (ChannelInputStream.java:74)
    at sun.nio.ch.ChannelInputStream.read (ChannelInputStream.java:103)
    at java.nio.file.Files.read (Files.java:3224)
    at java.nio.file.Files.readAllBytes (Files.java:3275)
    at org.apache.maven.buildcache.hash.SHA$Algorithm.hash (SHA.java:70)
    at org.apache.maven.buildcache.hash.HashAlgorithm.hash (HashAlgorithm.java:36)
    at org.apache.maven.buildcache.CacheControllerImpl.artifactDto (CacheControllerImpl.java:640)
    at org.apache.maven.buildcache.CacheControllerImpl.artifactDtos (CacheControllerImpl.java:628)
    at org.apache.maven.buildcache.CacheControllerImpl.save (CacheControllerImpl.java:508)
    at org.apache.maven.buildcache.BuildCacheMojosExecutionStrategy.execute (BuildCacheMojosExecutionStrategy.java:160)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:159)
...
```

We're using the build cache for a 1.7GB artifact. The build cache code currently reads all bytes of an artifact into a byte array, and computes the SHA-hash in one go: https://github.com/apache/maven-build-cache-extension/blob/b827d52410d96375e2e5ebafe3b95cf559649809/src/main/java/org/apache/maven/buildcache/hash/SHA.java#L70 .

For some hashes, there is a 'zero allocation hash' implementation: https://github.com/apache/maven-build-cache-extension/blob/b827d52410d96375e2e5ebafe3b95cf559649809/src/main/java/org/apache/maven/buildcache/hash/Zah.java#L101 . Unfortunately, there is no such version for SHA (all available hashes are defined [here](https://github.com/apache/maven-build-cache-extension/blob/b827d52410d96375e2e5ebafe3b95cf559649809/src/main/java/org/apache/maven/buildcache/hash/HashFactory.java#L40)).

Changing the SHA-implementation to use memory-mapping seems pretty straight forward, but changing the current behaviour might not be desirable. If I were to add a memory-mapped SHA implementation, would it be okay to add 4 SHA-entries to the HashFactory with `_MM` added to their name?

--- a/src/main/java/org/apache/maven/buildcache/hash/Hash.java
+++ b/src/main/java/org/apache/maven/buildcache/hash/Hash.java
@@ -57,4 +57,12 @@ public class Hash {
 
         Checksum checksum(int count);
     }
+
+    /**
+     * Memory policy
+     */
+    public enum MemoryPolicy {
+        Standard,
+        MemoryMappedBuffers
+    }
 }

--- a/src/main/java/org/apache/maven/buildcache/hash/HashFactory.java
+++ b/src/main/java/org/apache/maven/buildcache/hash/HashFactory.java
@@ -28,14 +28,18 @@ import net.openhft.hashing.LongHashFunction;
  * HashFactory
  */
 public enum HashFactory {
-    SHA1(new SHA("SHA-1")),
-    SHA256(new SHA("SHA-256")),
-    SHA384(new SHA("SHA-384")),
-    SHA512(new SHA("SHA-512")),
-    XX(new Zah("XX", LongHashFunction.xx(), Zah.MemoryPolicy.Standard)),
-    XXMM(new Zah("XXMM", LongHashFunction.xx(), Zah.MemoryPolicy.MemoryMappedBuffers)),
-    METRO(new Zah("METRO", LongHashFunction.metro(), Zah.MemoryPolicy.Standard)),
-    METRO_MM(new Zah("METRO+MM", LongHashFunction.metro(), Zah.MemoryPolicy.MemoryMappedBuffers));
+    SHA1(new SHA("SHA-1", Hash.MemoryPolicy.Standard)),
+    SHA1_MM(new SHA("SHA-1", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA256(new SHA("SHA-256", Hash.MemoryPolicy.Standard)),
+    SHA256_MM(new SHA("SHA-256", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA384(new SHA("SHA-384", Hash.MemoryPolicy.Standard)),
+    SHA384_MM(new SHA("SHA-384", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA512(new SHA("SHA-512", Hash.MemoryPolicy.Standard)),
+    SHA512_MM(new SHA("SHA-512", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    XX(new Zah("XX", LongHashFunction.xx(), Hash.MemoryPolicy.Standard)),
+    XXMM(new Zah("XXMM", LongHashFunction.xx(), Hash.MemoryPolicy.MemoryMappedBuffers)),
+    METRO(new Zah("METRO", LongHashFunction.metro(), Hash.MemoryPolicy.Standard)),
+    METRO_MM(new Zah("METRO+MM", LongHashFunction.metro(), Hash.MemoryPolicy.MemoryMappedBuffers));
 
     private static final Map<String, HashFactory> LOOKUP = new HashMap<>();
 

--- a/src/main/java/org/apache/maven/buildcache/hash/HashFactory.java
+++ b/src/main/java/org/apache/maven/buildcache/hash/HashFactory.java
@@ -29,13 +29,13 @@ import net.openhft.hashing.LongHashFunction;
  */
 public enum HashFactory {
     SHA1(new SHA("SHA-1", Hash.MemoryPolicy.Standard)),
-    SHA1_MM(new SHA("SHA-1", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA1_MM(new SHA("SHA-1+MM", Hash.MemoryPolicy.MemoryMappedBuffers)),
     SHA256(new SHA("SHA-256", Hash.MemoryPolicy.Standard)),
-    SHA256_MM(new SHA("SHA-256", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA256_MM(new SHA("SHA-256+MM", Hash.MemoryPolicy.MemoryMappedBuffers)),
     SHA384(new SHA("SHA-384", Hash.MemoryPolicy.Standard)),
-    SHA384_MM(new SHA("SHA-384", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA384_MM(new SHA("SHA-384+MM", Hash.MemoryPolicy.MemoryMappedBuffers)),
     SHA512(new SHA("SHA-512", Hash.MemoryPolicy.Standard)),
-    SHA512_MM(new SHA("SHA-512", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA512_MM(new SHA("SHA-512+MM", Hash.MemoryPolicy.MemoryMappedBuffers)),
     XX(new Zah("XX", LongHashFunction.xx(), Hash.MemoryPolicy.Standard)),
     XXMM(new Zah("XXMM", LongHashFunction.xx(), Hash.MemoryPolicy.MemoryMappedBuffers)),
     METRO(new Zah("METRO", LongHashFunction.metro(), Hash.MemoryPolicy.Standard)),

--- a/src/main/java/org/apache/maven/buildcache/hash/HashFactory.java
+++ b/src/main/java/org/apache/maven/buildcache/hash/HashFactory.java
@@ -28,14 +28,14 @@ import net.openhft.hashing.LongHashFunction;
  * HashFactory
  */
 public enum HashFactory {
-    SHA1(new SHA("SHA-1", Hash.MemoryPolicy.Standard)),
-    SHA1_MM(new SHA("SHA-1+MM", Hash.MemoryPolicy.MemoryMappedBuffers)),
-    SHA256(new SHA("SHA-256", Hash.MemoryPolicy.Standard)),
-    SHA256_MM(new SHA("SHA-256+MM", Hash.MemoryPolicy.MemoryMappedBuffers)),
-    SHA384(new SHA("SHA-384", Hash.MemoryPolicy.Standard)),
-    SHA384_MM(new SHA("SHA-384+MM", Hash.MemoryPolicy.MemoryMappedBuffers)),
-    SHA512(new SHA("SHA-512", Hash.MemoryPolicy.Standard)),
-    SHA512_MM(new SHA("SHA-512+MM", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA1(new SHA("SHA-1", "SHA-1", Hash.MemoryPolicy.Standard)),
+    SHA1_MM(new SHA("SHA-1+MM", "SHA-1", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA256(new SHA("SHA-256", "SHA-256", Hash.MemoryPolicy.Standard)),
+    SHA256_MM(new SHA("SHA-256+MM", "SHA-256", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA384(new SHA("SHA-384", "SHA-384", Hash.MemoryPolicy.Standard)),
+    SHA384_MM(new SHA("SHA-384+MM", "SHA-384", Hash.MemoryPolicy.MemoryMappedBuffers)),
+    SHA512(new SHA("SHA-512", "SHA-512", Hash.MemoryPolicy.Standard)),
+    SHA512_MM(new SHA("SHA-512+MM", "SHA-512", Hash.MemoryPolicy.MemoryMappedBuffers)),
     XX(new Zah("XX", LongHashFunction.xx(), Hash.MemoryPolicy.Standard)),
     XXMM(new Zah("XXMM", LongHashFunction.xx(), Hash.MemoryPolicy.MemoryMappedBuffers)),
     METRO(new Zah("METRO", LongHashFunction.metro(), Hash.MemoryPolicy.Standard)),

--- a/src/main/java/org/apache/maven/buildcache/hash/SHA.java
+++ b/src/main/java/org/apache/maven/buildcache/hash/SHA.java
@@ -35,17 +35,19 @@ public class SHA implements Hash.Factory {
     private static final ThreadLocal<MessageDigest> ALGORITHM = new ThreadLocal<>();
     private static final ThreadLocal<MessageDigest> CHECKSUM = new ThreadLocal<>();
 
+    private final String name;
     private final String algorithm;
     private final Hash.MemoryPolicy memoryPolicy;
 
-    SHA(String algorithm, Hash.MemoryPolicy memoryPolicy) {
+    SHA(String name, String algorithm, Hash.MemoryPolicy memoryPolicy) {
+        this.name = name;
         this.algorithm = algorithm;
         this.memoryPolicy = memoryPolicy != null ? memoryPolicy : Hash.MemoryPolicy.Standard;
     }
 
     @Override
     public String getAlgorithm() {
-        return algorithm;
+        return name;
     }
 
     @Override

--- a/src/main/java/org/apache/maven/buildcache/hash/SHA.java
+++ b/src/main/java/org/apache/maven/buildcache/hash/SHA.java
@@ -19,9 +19,13 @@
 package org.apache.maven.buildcache.hash;
 
 import java.io.IOException;
+import java.nio.channels.FileChannel;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.security.MessageDigest;
+
+import static java.nio.channels.FileChannel.MapMode.READ_ONLY;
+import static java.nio.file.StandardOpenOption.READ;
 
 /**
  * SHA
@@ -32,9 +36,11 @@ public class SHA implements Hash.Factory {
     private static final ThreadLocal<MessageDigest> CHECKSUM = new ThreadLocal<>();
 
     private final String algorithm;
+    private final Hash.MemoryPolicy memoryPolicy;
 
-    SHA(String algorithm) {
+    SHA(String algorithm, Hash.MemoryPolicy memoryPolicy) {
         this.algorithm = algorithm;
+        this.memoryPolicy = memoryPolicy != null ? memoryPolicy : Hash.MemoryPolicy.Standard;
     }
 
     @Override
@@ -44,7 +50,12 @@ public class SHA implements Hash.Factory {
 
     @Override
     public Hash.Algorithm algorithm() {
-        return new SHA.Algorithm(ThreadLocalDigest.get(ALGORITHM, algorithm));
+        switch (memoryPolicy) {
+            case MemoryMappedBuffers:
+                return new AlgorithmWithMM(ThreadLocalDigest.get(ALGORITHM, algorithm));
+            default:
+                return new SHA.Algorithm(ThreadLocalDigest.get(ALGORITHM, algorithm));
+        }
     }
 
     @Override
@@ -68,6 +79,28 @@ public class SHA implements Hash.Factory {
         @Override
         public byte[] hash(Path path) throws IOException {
             return hash(Files.readAllBytes(path));
+        }
+    }
+
+    private static class AlgorithmWithMM implements Hash.Algorithm {
+        private final MessageDigest digest;
+
+        private AlgorithmWithMM(MessageDigest digest) {
+            this.digest = digest;
+        }
+
+        @Override
+        public byte[] hash(byte[] array) {
+            return digest.digest(array);
+        }
+
+        @Override
+        public byte[] hash(Path path) throws IOException {
+            try (FileChannel channel = FileChannel.open(path, READ);
+                    CloseableBuffer buffer = CloseableBuffer.mappedBuffer(channel, READ_ONLY)) {
+                digest.update(buffer.getBuffer());
+                return digest.digest();
+            }
         }
     }
 

--- a/src/main/java/org/apache/maven/buildcache/hash/Zah.java
+++ b/src/main/java/org/apache/maven/buildcache/hash/Zah.java
@@ -34,19 +34,14 @@ import static java.nio.file.StandardOpenOption.READ;
  */
 public class Zah implements Hash.Factory {
 
-    public enum MemoryPolicy {
-        Standard,
-        MemoryMappedBuffers
-    }
-
     private final String name;
     private final LongHashFunction hash;
-    private final MemoryPolicy memoryPolicy;
+    private final Hash.MemoryPolicy memoryPolicy;
 
-    public Zah(String name, LongHashFunction hash, MemoryPolicy memoryPolicy) {
+    public Zah(String name, LongHashFunction hash, Hash.MemoryPolicy memoryPolicy) {
         this.name = name;
         this.hash = hash;
-        this.memoryPolicy = memoryPolicy != null ? memoryPolicy : MemoryPolicy.Standard;
+        this.memoryPolicy = memoryPolicy != null ? memoryPolicy : Hash.MemoryPolicy.Standard;
     }
 
     @Override

--- a/src/site/markdown/performance.md
+++ b/src/site/markdown/performance.md
@@ -28,23 +28,27 @@ performance tuning, measure results in relevant scenarios to validate results an
 By default, the cache uses the [XX](https://cyan4973.github.io/xxHash/) algorithm, which is a very fast
 non-cryptographic hash algorithm and is sufficient for most use cases. All supported values for `<hashAlgorithm>` are:
 
-| Identifier | Description                                                                                      |
-|------------|--------------------------------------------------------------------------------------------------|
-| `XX`       | **Default.** xxHash, non-cryptographic, fastest.                                                 |
-| `XXMM`     | xxHash with memory-mapped file I/O. Faster for large codebases; requires a JVM flag on JDK ≥ 17. |
-| `METRO`    | Metro hash, standard I/O.                                                                        |
-| `METRO+MM` | Metro hash with memory-mapped file I/O. Requires a JVM flag on JDK ≥ 17.                         |
-| `SHA-1`    | Cryptographic. Not recommended unless integrity guarantees are required.                         |
-| `SHA-256`  | Cryptographic.                                                                                   |
-| `SHA-384`  | Cryptographic.                                                                                   |
-| `SHA-512`  | Cryptographic, slowest.                                                                          |
+| Identifier   | Description                                                                                               |
+|--------------|-----------------------------------------------------------------------------------------------------------|
+| `XX`         | **Default.** xxHash, non-cryptographic, fastest.                                                          |
+| `XXMM`       | xxHash with memory-mapped file I/O. Faster for large codebases; requires a JVM flag on JDK ≥ 17.          |
+| `METRO`      | Metro hash, standard I/O.                                                                                 |
+| `METRO+MM`   | Metro hash with memory-mapped file I/O. Requires a JVM flag on JDK ≥ 17.                                  |
+| `SHA-1`      | Cryptographic. Not recommended unless integrity guarantees are required.                                  |
+| `SHA-1+MM`   | Cryptographic. Not recommended unless integrity guarantees are required. Requires a JVM flag on JDK ≥ 17. |
+| `SHA-256`    | Cryptographic.                                                                                            |
+| `SHA-256+MM` | Cryptographic. Requires a JVM flag on JDK ≥ 17.                                                           |
+| `SHA-384`    | Cryptographic.                                                                                            |
+| `SHA-384+MM` | Cryptographic. Requires a JVM flag on JDK ≥ 17.                                                           |
+| `SHA-512`    | Cryptographic, slowest.                                                                                   |
+| `SHA-512+MM` | Cryptographic, slowest. Requires a JVM flag on JDK ≥ 17.                                                  |
 
 ```xml
 
 <hashAlgorithm>XX</hashAlgorithm>
 ```
 
-The memory-mapped algorithms (`XXMM` and `METRO+MM`) require adding the following line to `.mvn/jvm.config` in the
+The memory-mapped algorithms (`XXMM` and those with the `+MM` suffix) require adding the following line to `.mvn/jvm.config` in the
 project root to run correctly on JDK ≥ 17:
 
 ```

--- a/src/test/java/org/apache/maven/buildcache/checksum/SHAHashTest.java
+++ b/src/test/java/org/apache/maven/buildcache/checksum/SHAHashTest.java
@@ -36,11 +36,16 @@ package org.apache.maven.buildcache.checksum;
  * specific language governing permissions and limitations
  * under the License.
  */
+import java.io.IOException;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
 
 import org.apache.maven.buildcache.hash.HashAlgorithm;
 import org.apache.maven.buildcache.hash.HashChecksum;
+import org.apache.maven.buildcache.hash.HashFactory;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
 
 import static org.apache.maven.buildcache.hash.HashFactory.SHA256;
 import static org.junit.jupiter.api.Assertions.assertEquals;
@@ -57,13 +62,17 @@ class SHAHashTest {
     private static final String FULL_CHECKSUM = "7305db9b2abccd706c256db3d97e5ff48d677cfe4d3a5904afb7da0e3950e1e2";
 
     private static final HashAlgorithm ALGORITHM = SHA256.createAlgorithm();
+    private static final HashAlgorithm ALGORITHM_MM = HashFactory.SHA256_MM.createAlgorithm();
+
     private static final HashChecksum CHECKSUM = SHA256.createChecksum(0);
+
+    @TempDir
+    Path tempDir;
 
     @Test
     void testEmptyArray() {
         byte[] emptyArray = new byte[0];
-        String hash = ALGORITHM.hash(emptyArray);
-        assertEquals(EMPTY_HASH, hash);
+        assertEquals(EMPTY_HASH, ALGORITHM.hash(emptyArray));
     }
 
     @Test
@@ -71,8 +80,17 @@ class SHAHashTest {
         String helloHash = ALGORITHM.hash(HELLO_ARRAY);
         assertEquals(HELLO_HASH, helloHash);
 
-        String worldHash = ALGORITHM.hash(WORLD_ARRAY);
-        assertEquals(WORLD_HASH, worldHash);
+        assertEquals(WORLD_HASH, ALGORITHM.hash(WORLD_ARRAY));
+        assertEquals(WORLD_HASH, ALGORITHM_MM.hash(WORLD_ARRAY));
+    }
+
+    @Test
+    void testFileHash() throws IOException {
+        Path testFile = tempDir.resolve("SHAHashTest.txt");
+        Files.write(testFile, HELLO_ARRAY);
+
+        assertEquals(HELLO_HASH, ALGORITHM.hash(testFile));
+        assertEquals(HELLO_HASH, ALGORITHM_MM.hash(testFile));
     }
 
     @Test

--- a/src/test/java/org/apache/maven/buildcache/hash/PerfTest.java
+++ b/src/test/java/org/apache/maven/buildcache/hash/PerfTest.java
@@ -83,6 +83,11 @@ public class PerfTest {
     }
 
     @Benchmark
+    public String sha256mm(HashState state) throws IOException {
+        return doTest(HashFactory.SHA256_MM, state);
+    }
+
+    @Benchmark
     public String xx(HashState state) throws IOException {
         return doTest(HashFactory.XX, state);
     }

